wild_lib/src/elf_writer.rs

```diff
@@ -282,9 +287,43 @@ impl SizedOutput {
             let mut section_buffers = split_output_into_sections(layout, &mut self.out);
             sort_eh_frame_hdr_entries(section_buffers.get_mut(output_section_id::EH_FRAME_HDR));
         }
+
+        if layout.args().build_id {
+            self.write_gnu_build_id_note(layout)?;
+        }
+        Ok(())
+    }
+
+    fn write_gnu_build_id_note<S: StorageModel>(&mut self, layout: &Layout<S>) -> Result {
+        let computed_build_id = self.compute_gnu_build_id_note();
+
+        let mut buffers = split_output_into_sections(layout, &mut self.out);
+        let e = LittleEndian;
+        let (note_header, mut rest) =
+            from_bytes_mut::<NoteHeader>(buffers.get_mut(output_section_id::NOTE_GNU_BUILD_ID))
+                .map_err(|_| anyhow!("Insufficient .note.gnu.build-id allocation"))?;
+        note_header.n_namesz.set(e, GNU_NOTE_NAME.len() as u32);
+        note_header.n_descsz.set(e, size_of::<GnuBuildId>() as u32);
+        note_header.n_type.set(e, NT_GNU_BUILD_ID);
+
+        let name_out = crate::slice::slice_take_prefix_mut(&mut rest, GNU_NOTE_NAME.len());
+        name_out.copy_from_slice(GNU_NOTE_NAME);
+
+        rest.copy_from_slice(&computed_build_id.as_slice());
+
         Ok(())
     }
 
+    #[instrument(skip_all, name = "Compute build ID")]
+    fn compute_gnu_build_id_note(&self) -> GnuBuildId {
+        let output_buffer = match &self.out {
+            OutputBuffer::Mmap(mmap) => mmap.deref(),
+            OutputBuffer::InMemory(vec) => vec.deref(),
+        };
+
+        GnuBuildId::new(Sha256::digest(output_buffer).into())
```

Comment:
Please use a faster hash algorithm (we can use BLAKE3 similar to Mold linker), and we can also make the hashing in parallel. Using the following code snippet, I reduced the time spent in hashing for building Clang from 1.5s to 100ms, which is about 2% of the total link time (w/ debug info). Pretty reasonable to me:

```rust
let mut hasher = blake3::Hasher::new();
        let chunks = output_buffer
            .chunks(bytesize::mib(16u64) as usize)
            .collect_vec();
        let hashes: Vec<_> = chunks.par_iter().map(|chunk| blake3::hash(chunk)).collect();
        for hash in hashes {
            hasher.update(hash.as_bytes());
        }

        GnuBuildId::new(hasher.finalize().into())
```